File: app/src/main/java/com/example/sachinsandbhor/retrofitsample/presentation/movieList/MovieListActivity.java
package com.example.sachinsandbhor.retrofitsample.presentation.movieList;

import android.content.Intent;
import android.databinding.DataBindingUtil;
import android.support.v7.app.AppCompatActivity;
import android.os.Bundle;
import android.support.v7.widget.LinearLayoutManager;
import android.view.View;
import android.widget.Toast;

import com.example.sachinsandbhor.retrofitsample.R;
import com.example.sachinsandbhor.retrofitsample.app.NetworkUtils;
import com.example.sachinsandbhor.retrofitsample.databinding.ActivityMainBinding;
import com.example.sachinsandbhor.retrofitsample.model.Movie;
import com.example.sachinsandbhor.retrofitsample.model.MovieResponse;
import com.example.sachinsandbhor.retrofitsample.presentation.movieDetail.DetailActivity;

import java.util.List;



public class MovieListActivity extends AppCompatActivity implements MovieListContract.MovieListView{

    public MovieListPresenter moviePresenter;

    private ActivityMainBinding activityMainBinding;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        activityMainBinding = DataBindingUtil.setContentView(this, R.layout.activity_main);
        activityMainBinding.recyclerview.setHasFixedSize(true);
        activityMainBinding.recyclerview.setLayoutManager(new LinearLayoutManager(this));
        moviePresenter = new MovieListPresenter(this);

    }

    @Override
    protected void onDestroy() {
        moviePresenter.detach();
        super.onDestroy();
    }

    @Override
    protected void onResume() {
        super.onResume();
        moviePresenter.setView(this);
        if(NetworkUtils.isNetworkAvaialble(this)) {
            moviePresenter.getMovieList();
        }else {
            onError("No internet connection");
        }
    }

    @Override
    public void setMovieList(MovieResponse movieResponse) {
        List<Movie> movies = movieResponse.getResults();
        activityMainBinding.recyclerview.setAdapter(new MovieListAdapter(movies, moviePresenter));
    }

    @Override
    public void startLoadingPost() {
        activityMainBinding.progressBar2.setVisibility(View.VISIBLE);
    }

    @Override
    public void stopLoadingPost() {
        activityMainBinding.progressBar2.setVisibility(View.GONE);
        activityMainBinding.recyclerview.setVisibility(View.VISIBLE);
    }

    @Override
    public void navigateToMovieDetail(int id) {
        Intent detailPage = new Intent(this, DetailActivity.class);
        detailPage.putExtra("ID", id);
        startActivity(detailPage);
    }

    @Override
    public void onError(String errorMessage) {
        Toast.makeText(this, errorMessage, Toast.LENGTH_SHORT).show();
    }
}

File: retrofitkotlin/src/main/java/com/sachinsandbhor/retrofitkotlin/presentation/movielist/MovieListActivity.kt
package com.sachinsandbhor.retrofitkotlin.presentation.movielist

import android.content.Intent
import android.databinding.DataBindingUtil
import android.os.Bundle
import android.support.v7.app.AppCompatActivity
import android.support.v7.widget.LinearLayoutManager
import android.support.v7.widget.RecyclerView
import android.view.View
import android.widget.Toast
import com.sachinsandbhor.retrofitkotlin.R
import com.sachinsandbhor.retrofitkotlin.app.EndlessRecyclerViewScrollListener
import com.sachinsandbhor.retrofitkotlin.databinding.ActivityMainBinding
import com.sachinsandbhor.retrofitkotlin.domain.MovieListResponse
import com.sachinsandbhor.retrofitkotlin.presentation.moviedetail.MovieDetailActivity


/**
 * Created by Sachin.Sandbhor on 07-03-2018.
 */

class MovieListActivity : AppCompatActivity(), MovieListContract.MovieListView{

    private lateinit var movieListDataBinding:ActivityMainBinding
    private lateinit var movieListPresenter :MovieListPresenter
    private lateinit var endlessRecyclerViewScrollListener:EndlessRecyclerViewScrollListener
    var pageNo:Int = 1
    lateinit var movieListAdapter:MovieListAdapter

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        movieListDataBinding = DataBindingUtil.setContentView(this, R.layout.activity_main)
        movieListPresenter = MovieListPresenter(this)
        setupRecyclerView()

    }

    private fun setupRecyclerView() {
        movieListAdapter = MovieListAdapter(movieListPresenter)
        var linearLayoutManager = LinearLayoutManager(this)
        movieListDataBinding.recyclerview.setHasFixedSize(true)
        movieListDataBinding.recyclerview.adapter = movieListAdapter
        movieListDataBinding.recyclerview.layoutManager = linearLayoutManager
        endlessRecyclerViewScrollListener = object : EndlessRecyclerViewScrollListener(linearLayoutManager) {
            override fun onLoadMore(page: Int, totalItemsCount: Int, view: RecyclerView?) {
                loadMoreData(page)
            }
        }
        movieListDataBinding.recyclerview.addOnScrollListener(endlessRecyclerViewScrollListener)
    }

    private fun loadMoreData(page: Int) {
        pageNo = pageNo+1
        movieListPresenter.getMovieList(pageNo)
    }

    override fun setMovie(movieListResponse: MovieListResponse) {
        if(movieListResponse.page > 1 ){
            movieListAdapter.addAllData(movieListResponse)
        }else {
            movieListDataBinding.recyclerview.visibility = View.VISIBLE
            movieListAdapter.movieList.clear()
            movieListAdapter.addAllData(movieListResponse)
        }
    }

    override fun startLoading() {
        movieListDataBinding.progressBar.visibility = View.VISIBLE
    }

    override fun stopLoading() {
        movieListDataBinding.progressBar.visibility = View.INVISIBLE
    }

    override fun onResume() {
        super.onResume()
        movieListPresenter.setView(this)
        movieListPresenter.getMovieList(1)
    }

    override fun onError(message: String?) {
        Toast.makeText(this, message, Toast.LENGTH_LONG).show()
    }

    override fun navigateUser(id: Int) {
        val detailPage = Intent(this, MovieDetailActivity::class.java)
        detailPage.putExtra("ID", id)
        startActivity(detailPage)
    }

}
